Get embedding working through conda

kmod opened the pull request:
I just copied the conda-forge approach: they have a libpython_static package that just contains the libpython.a library. I think this conflicts with the python package which also contains this file, but it seems to work?

I also made a couple other small changes. The uwsgi feedstock tests pass, and the conda package works enough to spawn a uwsgi process that exits immediately, but we'll need a bunch more packages to run our uwsgi benchmark.

kmod committed:
create a libpython-static library

I think the actual contents are superfluous because they're provided by
the base pyston2.3 package, but the package is still needed for things
that explicitly depend on libpython-static (such as uwsgi)

kmod committed:
Change build_pkgs to use conda-forge instead of anaconda

I ran into some GLIBC version errors when building uwsgi, and then switched all
of the commands to only use conda-forge and it worked

undingen commented:
Seeing a new error with this on my machine :/ currently rerunning
```
BOLT-INFO: SCTC: patched 141 tail calls (141 forward) tail calls (0 backward) from a total of 141 while removing 16 double jumps and removing 122 basic blocks totalling 610 bytes of code. CTCs total execution count is 67918035 and the number of times CTCs are taken is 10699885.
BOLT-INFO: Peephole: 0 instructions shortened.
BOLT-INFO: Peephole: 651 double jumps patched.
BOLT-INFO: Peephole: 195 tail call traps inserted.
BOLT-INFO: Peephole: 0 useless conditional branches removed.
BOLT-INFO: padding code to 0xc00000 to accommodate hot text
Unsupported DWARFLocationEntry Kind.
UNREACHABLE executed at /opt/conda/conda-bld/bolt_1635331680196/work/bolt/lib/Rewrite/DWARFRewriter.cpp:447!
 #0 0x0000559442c0b3c0 PrintStackTraceSignalHandler(void*) (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x7a13c0)
 #1 0x0000559442c0906e SignalHandler(int) (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x79f06e)
 #2 0x00007fac450f6730 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12730)
 #3 0x00007fac44bad7bb raise (/lib/x86_64-linux-gnu/libc.so.6+0x377bb)
 #4 0x00007fac44b98535 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22535)
 #5 0x0000559442b9ff1a (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x735f1a)
 #6 0x0000559442aee7ec bool llvm::function_ref<bool (llvm::DWARFLocationEntry const&)>::callback_fn<llvm::bolt::DWARFRewriter::updateUnitDebugInfo(unsigned long, llvm::DWARFUnit&, llvm::bolt::SimpleBinaryPatcher&, llvm::bolt::DebugAbbrevWriter&, llvm::Optional<unsigned long>)::'lambda'(llvm::DWARFLocationEntry const&)>(long, llvm::DWARFLocationEntry const&) (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x6847ec)
 #7 0x00005594439f4aa3 llvm::DWARFDebugLoclists::visitLocationList(unsigned long*, llvm::function_ref<bool (llvm::DWARFLocationEntry const&)>) const (.localalias) (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x158aaa3)
 #8 0x0000559442af3522 llvm::bolt::DWARFRewriter::updateUnitDebugInfo(unsigned long, llvm::DWARFUnit&, llvm::bolt::SimpleBinaryPatcher&, llvm::bolt::DebugAbbrevWriter&, llvm::Optional<unsigned long>) (.localalias) (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x689522)
 #9 0x0000559442af5259 llvm::bolt::DWARFRewriter::updateDebugInfo()::'lambda0'(unsigned long, llvm::DWARFUnit*)::operator()(unsigned long, llvm::DWARFUnit*) const (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x68b259)
#10 0x0000559442af879d llvm::bolt::DWARFRewriter::updateDebugInfo() (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x68e79d)
#11 0x0000559442aba5b2 llvm::bolt::RewriteInstance::updateMetadata() (.localalias) (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x6505b2)
#12 0x0000559442ae6076 llvm::bolt::RewriteInstance::run() (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x67c076)
#13 0x000055944264d5f9 main (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x1e35f9)
#14 0x00007fac44b9a09b __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2409b)
#15 0x00005594426a96d2 _start (/opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt+0x23f6d2)
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /opt/conda/conda-bld/pyston2.3-split_1635784395286/_build_env/bin/llvm-bolt build/release_install/opt/conda/conda-bld/pyston2.3-split_1635784395286/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/python3 -o build/release_install/opt/conda/conda-bld/pyston2.3-split_1635784395286/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/python3.bolt -data=build/release_install/opt/conda/conda-bld/pyston2.3-split_1635784395286/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/python3.bolt.fdata -update-debug-sections -reorder-blocks=cache+ -reorder-functions=hfsort+ -split-functions=3 -icf=1 -inline-all -split-eh -reorder-functions-use-hot-size -peepholes=all -jump-tables=aggressive -inline-ap -indirect-call-promotion=all -dyno-stats -use-gnu-stack
make: *** [Makefile:248: build/release_install/opt/conda/conda-bld/pyston2.3-split_1635784395286/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/python3.bolt] Aborted (core dumped)

```

kmod committed:
Pin our dependencies in the same way the conda-forge python feedstock does

More importantly than setting the version, pinning them in this way
automatically adds them as run dependencies.

This also sets openssl to v1.1.1 which is important because it looks
like the conda-forge ecosystem is built on that version.
It also sets the compiler version which should avoid some other problems.

undingen reviewed (changes requested):
We have to add `-g -gdwarf-4` to `C{,XX}FLAGS_NODIST` to get it to build with gcc11 (I have a fix for this). But even with this `pyston/conda/build_pkgs.sh` is failing with a segfault while installing pillow :/

undingen reviewed (changes requested):
also I think it's better to remove `conda index /conda_pkgs` and `-c /conda_pkgs` arg and replace it with `--use-local`. Because the conda index call currently sets wrong directory ownership so needs to be erased by root user which is annoying and `--use-local` also works if one manually continues a previously broken build which did not copy anything into `/conda_pkgs` but some packages build correctly.

kmod commented:
I don't understand this second comment, how does it know to pick up packages from /conda_pkgs? Or are you saying it'll pick up packages if you docker exec into the existing container and then run commands? My goal was to make it so that you don't have to build bolt every time, including if you change the build_pkgs script and want to rerun it.

undingen commented:
I had in mind what I implemented here https://github.com/pyston/pyston/pull/122/commits/7f5f17737721aae2895d783be29a59d0045a1507 . I think with this if pyston packaging fails one can manually start the stopped docker container and run docker exec to get a shell and rerun the failed package build and it will reuse the already build bolt packages. Maybe I misunderstand why you added the `conda index /conda_pkgs` code and we still need it.

kmod commented:
Ah ok, I added the /conda_pkgs thing so that you could rerun build_pkgs and it wouldn't rebuild bolt. Now that it's getting it from the pyston channel that solves the problem as well, so I'll remove it.

Not sure what's causing the segfault yet, I'll look into it.

undingen commented:
I tested your improved patch now more than 5 times on the github and locally and it never failed for me. I assume that the crash was caused by gcc11 and now that we pin version 9 it does not happen anymore. 
Removing `-m variants.yaml` made it crash again at least once on the CI.
So I think the issue has nothing todo with your patch and we have to investigate it when switching to gcc-11.

LGTM I merged it in with a small change which removes: `-c /conda_pkgs` from `conda install patch -y -c /conda_pkgs`

kmod commented:
awesome thanks!
